### PR TITLE
ksmtuned: New test case to check kSM works well

### PR DESCRIPTION
**Pull request by vivianQizhu**
id: 1899901
Signed-off-by: Qianqian Zhu <qizhu@redhat.com>

### CODE REVIEW

**Comment by vivianQizhu** [comment]
@huangyum Please help review. I didn't add Windows part yet, please help check the process first, if the step is okay I will add Windows stress part then, thanks.

**Comment by vivianQizhu** [comment]
(1/2) type_specific.io-github-autotest-qemu.ksmtuned: PASS (164.08 s)

File: qemu/tests/ksmtuned.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                                     (int(params['mem']) // 2))
+            stress_test = VMStress(vm, "stress", params)
+            stress_test.load_stress_tool()
+            time.sleep(10)
+        qemu_pid = vm.process.get_pid()
+        qemu_used_mem = int(process.getoutput(params['cmd_get_qemu_used_mem'] %
+                                              qemu_pid, shell=True)) * 4 * 1024
+        free_mem_host = utils_memory.freememtotal()
+        vm.destroy()
+        ksm_status = process.getoutput(params['cmd_check_ksm_status'])
```
Comment (huangyum) [inline qemu/tests/ksmtuned.py:54]: Let's get ksm status before vm.destroy.

File: qemu/tests/ksmtuned.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                (qemu_used_mem + ksm_thres) <= total_mem_host):
+            if ksm_status != '0':
+                test.fail('Ksm should not start.')
+            if stress:
+                test.error('The host resource is not consumed as expected.')
+        elif ksm_status == '0':
+            test.fail('Ksm should start but it does not.')
+
+    total_mem_host = utils_memory.memtotal()
+    free_mem_host = utils_memory.freememtotal()
```
Comment (huangyum) [inline qemu/tests/ksmtuned.py:68]: Had better drop caches first before get free memory.

File: qemu/tests/ksmtuned.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    ksm_thres = process.getoutput(params['cmd_get_thres'], shell=True)
+    ksm_thres = int(total_mem_host *
+                    (int(re.findall('\\d+', ksm_thres)[0]) / 100))
+    guest_mem = (free_mem_host - ksm_thres) // 2
+    check_ksm(guest_mem)
```
Comment (huangyum) [inline qemu/tests/ksmtuned.py:73]: Suggest to check ksmtuned service status first, start it if it's disabled, and restore it in finally.

File: qemu/tests/ksmtuned.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    ksm_config_file = params['ksm_config_file']
+    backup_file = ksm_config_file + '.backup'
+    copyfile(ksm_config_file, backup_file)
+    threshold = 60
+    with open(ksm_config_file, "a+") as f:
+        f.write('%s=%s' % (params['ksm_thres_conf'], threshold))
+    process.run(params['cmd_restart_ksmtuned'])
+    ksm_thres = total_mem_host * (threshold / 100)
+    try:
+        check_ksm((guest_mem * 2), stress=True)
```
Comment (huangyum) [inline qemu/tests/ksmtuned.py:84]: Had better re-calculate guest memory, e.g. guest_mem = total_mem_host - ksm_thres/2.

File: qemu/tests/ksmtuned.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        vm = env.get_vm(vm_name)
+        vm.verify_alive()
+        if stress:
+            params['stress_args'] = ('--cpu 4 --io 4 --vm 2 --vm-bytes %sM' %
+                                     (int(params['mem']) // 2))
+            stress_test = VMStress(vm, "stress", params)
+            stress_test.load_stress_tool()
+            time.sleep(10)
+        qemu_pid = vm.process.get_pid()
+        qemu_used_mem = int(process.getoutput(params['cmd_get_qemu_used_mem'] %
```
Comment (huangyum) [inline qemu/tests/ksmtuned.py:50]: I guess we need wait guest boot up, then qemu used memory will be stable.

**Comment by vivianQizhu** [comment]
@huangyum All updated, but with this modification the ksm does not start as expected after load stress on guest, should we wait a while somewhere?

**Comment by vivianQizhu** [comment]
22:13:37 INFO | The ksm threshold is 9699405.6, the memory allocated by qemu is 3031040, and the total free memory on host is 3362132.
22:13:37 INFO | Running 'systemctl restart ksmtuned.service'
22:13:37 INFO | Command 'systemctl restart ksmtuned.service' finished with 0 after 0.01341557502746582s
22:13:37 ERROR|
22:13:37 ERROR| Reproduced traceback from: /usr/local/lib/python3.6/site-packages/avocado_framework_plugin_vt-83.0-py3.6.egg/avocado_vt/test.py:442
22:13:37 ERROR| Traceback (most recent call last):
22:13:37 ERROR|   File "/var/lib/avocado/data/avocado-vt/virttest/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/ksmtuned.py", line 88, in run
22:13:37 ERROR|     check_ksm((total_mem_host - ksm_thres//2), stress=True)
22:13:37 ERROR|   File "/var/lib/avocado/data/avocado-vt/virttest/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/ksmtuned.py", line 65, in check_ksm
22:13:37 ERROR|     test.fail('Ksm should start but it does not.')
22:13:37 ERROR|   File "/usr/local/lib/python3.6/site-packages/avocado_framework-83.0-py3.6.egg/avocado/core/test.py", line 953, in fail
22:13:37 ERROR|     raise exceptions.TestFail(message)
22:13:37 ERROR| avocado.core.exceptions.TestFail: Ksm should start but it does not.

File: qemu/tests/ksmtuned.py (hunk 1 above)
Comment (vivianQizhu) [reply qemu/tests/ksmtuned.py:53]: @huangyum  I don't know why but it can only passed with getting ksm status after vm.destroy.

Comment (vivianQizhu) [reply qemu/tests/ksmtuned.py:53]: Extend the sleeping after loading stress to 30s, it passed, not sure if this long is proper.

**Comment by huangyum** [comment]
@vivianQizhu It passed on my host. However, two cases are generated.
 (1/2) type_specific.io-github-autotest-qemu.ksmtuned: PASS (106.15 s)
 (2/2) io-github-autotest-qemu.ksm_services.ksmtuned: PASS (2.14 s)

File: qemu/tests/ksmtuned.py (hunk 1 above)
Comment (huangyum) [reply qemu/tests/ksmtuned.py:53]: 30s looks good to me.

**Comment by vivianQizhu** [comment]
> @vivianQizhu It passed on my host. However, two cases are generated.
> (1/2) type_specific.io-github-autotest-qemu.ksmtuned: PASS (106.15 s)
> (2/2) io-github-autotest-qemu.ksm_services.ksmtuned: PASS (2.14 s)

Ah, seems now we can not have the case name duplicated with the variant name in other test case, let me change a different name for it.

**Comment by vivianQizhu** [comment]
@huangyum Update test case name, please help check. Thanks.
 (1/1) type_specific.io-github-autotest-qemu.ksmtuned_threshold: PASS (221.22 s)

**Comment by vivianQizhu** [comment]
@huangyum Any more comments? Thanks.

**Review by huangyum — COMMENTED** [review]
Sorry, seems I added review comments before but forgot to submit.

**Comment by maxujun** [comment]
Failed to run on ppc platform due to "qemu-kvm: Memory size 0x1836300000 is not aligned to 256 MiB",seems the mem size of qemu cmdline "-m 99171 " have problem on ppc.

**Comment by vivianQizhu** [comment]
> Failed to run on ppc platform due to "qemu-kvm: Memory size 0x1836300000 is not aligned to 256 MiB",seems the mem size of qemu cmdline "-m 99171 " have problem on ppc.

@maxujun Would you help provide more information, like what value is okay on ppc platform? Thanks.

File: qemu/tests/ksmtuned_threshold.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        elif ksm_status == '0':
+            test.fail('Ksm should start but it does not.')
+
+    total_mem_host = utils_memory.memtotal()
+    utils_memory.drop_caches()
+    free_mem_host = utils_memory.freememtotal()
+    ksm_thres = process.getoutput(params['cmd_get_thres'], shell=True)
+    ksm_thres = int(total_mem_host *
+                    (int(re.findall('\\d+', ksm_thres)[0]) / 100))
+    guest_mem = (free_mem_host - ksm_thres) // 2
```
Comment (vivianQizhu) [inline qemu/tests/ksmtuned_threshold.py:73]: @maxujun Please help check this policy, that's where the guest memory comes from, should it be different on ppc platform?

Comment (maxujun) [reply qemu/tests/ksmtuned_threshold.py]: guest_mem = guest_mem - guest_mem%256

Comment (maxujun) [reply qemu/tests/ksmtuned_threshold.py]: Should be:guest_mem = guest_mem - guest_mem%(256*1024)

Have problem after that.
2020-12-15 03:23:34,126 stacktrace       L0045 ERROR|   File "/var/lib/avocado/data/avocado-vt/virttest/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/ksm_ksmtuned.py", line 91, in run
2020-12-15 03:23:34,126 stacktrace       L0045 ERROR|     check_ksm(guest_mem)
2020-12-15 03:23:34,126 stacktrace       L0045 ERROR|   File "/var/lib/avocado/data/avocado-vt/virttest/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/ksm_ksmtuned.py", line 60, in check_ksm
2020-12-15 03:23:34,126 stacktrace       L0045 ERROR|     (qemu_used_mem + ksm_thres) <= total_mem_host):
2020-12-15 03:23:34,127 stacktrace       L0045 ERROR| TypeError: must be str, not int

Comment (vivianQizhu) [reply qemu/tests/ksmtuned_threshold.py]: Thanks @maxujun That's introduced by the latest change, I've modified it, and add the `guest_mem` adjustment for ppc platform, please help check it again.

Comment (MiriamDeng) [reply qemu/tests/ksmtuned_threshold.py]: By the way, we need to set over 2G memory size for each guest at least on ppc. Thanks

File: qemu/tests/ksm_ksmtuned.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    ksm_config_file = params['ksm_config_file']
+    backup_file = ksm_config_file + '.backup'
+    copyfile(ksm_config_file, backup_file)
+    threshold = 60
+    with open(ksm_config_file, "a+") as f:
+        f.write('%s=%s' % (params['ksm_thres_conf'], threshold))
+    process.run(params['cmd_restart_ksmtuned'])
+    ksm_thres = total_mem_host * (threshold / 100)
+    try:
+        check_ksm((total_mem_host - ksm_thres//2), stress=True)
```
Comment (maxujun) [inline qemu/tests/ksm_ksmtuned.py:94]: Failed to boot up guest here with "qemu-kvm: Memory size 0x2c03c00000 is not aligned to 256 MiB"

Comment (vivianQizhu) [reply qemu/tests/ksm_ksmtuned.py]: @maxujun Updated, please help re-review, thanks.

File: qemu/tests/ksm_ksmtuned.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            params['cmd_get_qemu_used_mem'] % qemu_pid, shell=True) + 'K', 'B')
+        qemu_used_mem = int(float(qemu_used_page)) * 4
+        free_mem_host = utils_memory.freememtotal()
+        ksm_status = process.getoutput(params['cmd_check_ksm_status'])
+        vm.destroy()
+        logging.info('The ksm threshold is %s, the memory allocated by qemu is'
+                     ' %s, and the total free memory on host is %s.' %
+                     (ksm_thres, qemu_used_mem, free_mem_host))
+        if (free_mem_host >= ksm_thres and
+                (qemu_used_mem + ksm_thres) <= total_mem_host):
```
Comment (huangyum) [inline qemu/tests/ksm_ksmtuned.py:62]: Pls remove this condition `(qemu_used_mem + ksm_thres) <= total_mem_host`. It's possible that other process consumes memory causing ksm start.

File: qemu/tests/ksm_ksmtuned.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        the ksm status in host. Ksm should not start in the host
+    6. Repeat step 2~5 under it broke the rule in step 5
+
+    :param test: kvm test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    def check_ksm(mem, stress=False):
+        """
+        :param mem: Boot guest with given memory
```
[the remark this answers is not in the document]
Comment (vivianQizhu) [reply qemu/tests/ksm_ksmtuned.py]: @huangyum Good suggestion, added, thanks.

File: qemu/tests/ksm_ksmtuned.py (hunk 8 above)
Comment (vivianQizhu) [reply qemu/tests/ksm_ksmtuned.py]: @huangyum Okay, updated. Please help re-review, thanks.

**Review by huangyum — COMMENTED** [review]
(1/1) type_specific.io-github-autotest-qemu.ksm_ksmtuned: PASS (142.94 s)
ACK

**Comment by maxujun** [comment]
(1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: FAIL: Ksm should start but it does not. 
Failed on ppc platform.I'm not sure if it's a product bug or a case design issue.
@MiriamDeng Could you help check it?

**Comment by MiriamDeng** [comment]
Hi maxujun,
I guess you didn't set it in the configuration file before your test,But even if I set it , there's failures that need to be checked
#  cat /etc/ksmtuned.conf 
KSM_THRES_COEF=20
05:09:47 ERROR| Reproduced traceback from: /home/kar/workspace/avocado/avocado/core/test.py:767
05:09:47 ERROR| Traceback (most recent call last):
05:09:47 ERROR|   File "/home/kar/workspace/avocado-vt/avocado_vt/test.py", line 309, in runTest
05:09:47 ERROR|     raise self.__status  # pylint: disable=E0702
05:09:47 ERROR|   File "/home/kar/workspace/avocado-vt/avocado_vt/test.py", line 255, in setUp
05:09:47 ERROR|     self._runTest()
05:09:47 ERROR|   File "/home/kar/workspace/avocado-vt/avocado_vt/test.py", line 390, in _runTest
05:09:47 ERROR|     run_func(self, params, env)
05:09:47 ERROR|   File "/var/lib/avocado/data/avocado-vt/virttest/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/ksm_ksmtuned.py", line 96, in run
05:09:47 ERROR|     check_ksm(guest_mem, stress=True)
05:09:47 ERROR|   File "/var/lib/avocado/data/avocado-vt/virttest/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/ksm_ksmtuned.py", line 53, in check_ksm
05:09:47 ERROR|     params['cmd_get_qemu_used_mem'] % qemu_pid, shell=True) + 'K', 'B')
05:09:47 ERROR|   File "/home/kar/workspace/avocado-vt/virttest/utils_misc.py", line 2369, in normalize_data_size
05:09:47 ERROR|     value = match.group(1)
05:09:47 ERROR| AttributeError: 'NoneType' object has no attribute 'group'

@vivianQizhu , could you help to check it ?  Thanks a lot.

**Comment by vivianQizhu** [comment]
@MiriamDeng Thanks for your review. That's not about pre-configuration, I have run test on @maxujun 's host, checking the log it seems previous steps all works well, but ksm does not start set after stress the guest, I guest the trigger condition should be different on ppc platform(or on different environment). And the issue you hit seems like the qemu process is killed during the test, did you run other test at the same time when this test is ongoing?

**Comment by vivianQizhu** [comment]
@MiriamDeng @maxujun Can we tell something from this log: `00:55:09 INFO | The ksm threshold is 319148428.8, the memory allocated by qemu is 23569559552, and the total free memory on host is 141911040.`? And would you help introduce how do you test this case usually on powerpc? Is that very same with the steps I write in this case?

**Comment by MiriamDeng** [comment]
@vivianQizhu, I took pto in the last week and will provide the parameters on ppc soon, thanks.

File: qemu/tests/ksm_ksmtuned.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        if stress:
+            params['stress_args'] = ('--cpu 4 --io 4 --vm 2 --vm-bytes %sM' %
+                                     (int(params['mem']) // 2))
+            stress_test = VMStress(vm, "stress", params)
+            stress_test.load_stress_tool()
+            time.sleep(30)
+        qemu_pid = vm.get_pid()
+        qemu_used_page = utils_misc.normalize_data_size(process.getoutput(
+            params['cmd_get_qemu_used_mem'] % qemu_pid, shell=True) + 'K', 'B')
+        qemu_used_mem = int(float(qemu_used_page)) * 4
```
Comment (MiriamDeng) [inline qemu/tests/ksm_ksmtuned.py:54]: 4K is for x86 but for ppc, it should be 64k
[root@ibm-p9b-19 kar]# getconf PAGE_SIZE
65536

Comment (vivianQizhu) [reply qemu/tests/ksm_ksmtuned.py]: @MiriamDeng Thanks a lot. Updated, please help re-review.

File: qemu/tests/ksm_ksmtuned.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    status_ksm_service = process.system(
+        params['cmd_status_ksmtuned'], ignore_status=True)
+    if status_ksm_service != 0:
+        process.run(params['cmd_start_ksmtuned'])
+    check_ksm(guest_mem)
+
+    ksm_config_file = params['ksm_config_file']
+    backup_file = ksm_config_file + '.backup'
+    copyfile(ksm_config_file, backup_file)
+    threshold = 60
```
Comment (MiriamDeng) [inline qemu/tests/ksm_ksmtuned.py:88]: Can we incease threshold a bit for ppc, the service will be triggered while threshold is 70 and 80, thanks.

Comment (MiriamDeng) [reply qemu/tests/ksm_ksmtuned.py:88]: Prefer to use a parameter for threshold for different architectures. Thanks

Comment (vivianQizhu) [reply qemu/tests/ksm_ksmtuned.py]: @MiriamDeng Yes, updated, please help check, thanks.

**Comment by MiriamDeng** [comment]
There's minor issue,
03:49:13 ERROR|     ksm_thres = total_mem_host * (threshold / 100)
03:49:13 ERROR| TypeError: unsupported operand type(s) for /: 'str' and 'int'

**Comment by MiriamDeng** [comment]
python3 ConfigTest.py  --guestname=RHEL.8.4.0 --platform=ppc64le --driveformat=virtio_scsi --nicmodel=virtio_net --mem=262144 --vcpu=8 --testcase=ksm_ksmtuned --customsparam="machine_type_extra_params = max-cpu-compat=power8" --clone=no
 (1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.ksm_ksmtuned: PASS (292.38 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-01-05T04.32-4a7856b/results.html
JOB TIME   : 293.81 s

**Comment by vivianQizhu** [comment]
Thanks @huangyum @maxujun @MiriamDeng

**Pull request merged**